### PR TITLE
chore: Remove dependency of AWS.Lambda.Powertools.Common on all Utilities

### PR DESCRIPTION
**Pull request by hjgraca**
> Please provide the issue number

Issue number: #339 

## Summary

Utilities have a dependency on the AWS.Lambda.Powertools.Common project.
When the utility is compiled and packaged into a nuget package the AWS.Lambda.Powertools.Common is a dependency.
This behaviour is causing [Diamond dependency issues](https://learn.microsoft.com/en-us/dotnet/standard/library-guidance/dependencies) in the project.
To address this we should copy the required files to each utility and make them completely isolated. This allows the client to not have to worry about dependencies between utilities and AWS.Lambda.Powertools.Common.
There is no need to delete the AWS.Lambda.Powertools.Common project but link the files in other projects, this will make it more readable in the solution explorer and makes maintenance easier because it's all in a single project.

### Changes

- Remove dependency of `AWS.Lambda.Powertools.Common` project in all utilities
- Create a Compile Include Link from `AWS.Lambda.Powertools.Common` files to all utilities csproj
Pasted image 20230704131824
- Add a reference to `AspectInjector`
Pasted image 20230704115449
- This will result in linked files in the `Common` folder of the utility project
image

Updated test project dependencies to latest versions

> Please provide a summary of what's being changed

### User experience

> Please share what the user experience looks like before and after this change

## Checklist

Please leave checklist items unchecked if they do not apply to your change.

* [x] [Meets tenets criteria](https://docs.powertools.aws.dev/lambda/dotnet/tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-dotnet/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by hjgraca** [comment]
After some conversations with @amirkaws and his feedback, I have made some changes to the implementation.

### TLDR:

- Add `Directory.Build.props` and `Directory.Build.targets` [more info](https://learn.microsoft.com/en-us/visualstudio/msbuild/customize-by-directory?view=vs-2022#directorybuildprops-and-directorybuildtargets)
	- These two files allows us to modify all csproj files in the solution or where the files are placed in the folder structure
	- Those files were added to the `src` folder. They modify all csproj in that folder where all utilities are. [Search scope](https://learn.microsoft.com/en-us/visualstudio/msbuild/customize-by-directory?view=vs-2022#search-scope)
- Add Central Package Management [more info](https://learn.microsoft.com/en-us/nuget/consume-packages/central-package-management)
	- Dependency management is a core feature of NuGet. Managing dependencies for a single project can be easy. Managing dependencies for multi-project solutions can prove to be difficult as they start to scale in size and complexity. In situations where you manage common dependencies for many different projects, you can leverage NuGet's central package management (CPM) features to do all of this from the ease of a single location.	
	- Add a file `Directory.Packages.props` to the src and test folders. This allows us to manage nuget package versions for all projects in a central location. I separated src and tests because tests have many dependencies and can cause confusion. [Search scope info](https://learn.microsoft.com/en-us/nuget/consume-packages/central-package-management#central-package-management-rules)

### Implementation

#### Part 1 `Directory.Build.props`

- Use MSbuild to:
	- Manage all common properties across all projects
	- These properties can be overridden by the project if it wishes to do so
image


#### Part 2 `Directory.Build.targets` (**changes only happen in Release configuration**)

- Use MSbuild to:
	- Copy/Link `AWS.Lambda.Powertools.Common` *.cs files to the destination project
	- Remove depdendency of Common project by removing `<ProjectReference Remove="..\AWS.Lambda.Powertools.Common\AWS.Lambda.Powertools.Common.csproj" />`
	- Add PackageReference `<PackageReference Include="AspectInjector" />`
- This transformation only happens in **Release** configuration

image

Once in `Release` configuration or when running `dotnet build -c Release` the Common folder is added to the project

image

#### Part 3 Central Package Management

[Docs](https://learn.microsoft.com/en-us/nuget/consume-packages/central-package-management)

Starting with [NuGet 6.2](https://learn.microsoft.com/en-us/nuget/release-notes/nuget-6.2), you can centrally manage your dependencies in your projects with the addition of a Directory.Packages.props file and an MSBuild property.

This means that the versions of the referenced packages will be defined in the central `Directory.Packages.props` file

File contents:
image

In the csproj files of the projects the referenced packages will not have a version

image

If needed individual projects can [override the version](https://learn.microsoft.com/en-us/nuget/consume-packages/central-package-management#overriding-package-versions)

#### Part 4 Nuget package contents

The resulting **Release** package will have no `AWS.Lambda.Powertools.Common` reference and will include `AspectInjector` nuget package. This will make the utility completely independent from `AWS.Lambda.Powertools.Common` Nuget package at runtime

**Old:**

image


**New:**

image

**Comment by hjgraca** [comment]
Also updated nuget packages. Once merged will close dependabot pull requests.

**Comment by hjgraca** [comment]
Removed `AWS.Lambda.Powertools.Common` project reference from test projects.
This reference now comes from `Directory.Build.props`. This file was added to test folder.

Contents:
carbon (1)

**Review by amirkaws — APPROVED** [review]
LGTM
